File: python/20. Valid Parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        dic = {'(':')','[':']','{':'}'}
        left = ['(','[','{']
        right = [')',']','}']
        stack = ''
        for item in s:
            if item in left:
                stack+=item
                continue
            if item in right:
                if len(stack)==0:
                    return False
                if dic[stack[-1]]==item:
                    stack=stack[:-1]
                else:
                    return False
                    
        if len(stack)!=0:
            return False
        return True

File: c/20. Valid Parentheses.cpp
//
// Created by root on 9/22/16.
//

#include <map>
#include <stack>
#include <iostream>

using namespace std;
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        map<char,char> dic={{')','('},{']','['},{'}','{'}};
        for(auto& it: s ){
            const auto & p = dic.find(it);
            if(p!=dic.end()){
                if(st.empty() || p->second!=st.top())
                    return false;
                st.pop();
            }
            else
                st.push(it);
        }

        return st.empty();



    }
};
int main(){
    Solution s;
    bool re = s.isValid("()");
    cout<< re << endl;
    re = s.isValid("{()}");
    cout<< re << endl;
    re = s.isValid("{{{[[[}}}");
    cout<< re << endl;


}